lapack/tpls/KokkosLapack_gesv_tpl_spec_decl.hpp

```diff
@@ -390,6 +390,151 @@ KOKKOSLAPACK_CGESV_MAGMA(Kokkos::LayoutLeft, Kokkos::CudaSpace, false)
 }  // namespace KokkosLapack
 #endif  // KOKKOSKERNELS_ENABLE_TPL_MAGMA
 
+// ROCSOLVER
+#ifdef KOKKOSKERNELS_ENABLE_TPL_CUSOLVER
+#include "KokkosLapack_cusolver.hpp"
+
+namespace KokkosLapack {
+namespace Impl {
+
+template <class ExecutionSpace, class IPIVViewType, class AViewType,
+          class BViewType>
+void cusolverGesvWrapper(const ExecutionSpace& space, const IPIVViewType& IPIV,
+                         const AViewType& A, const BViewType& B) {
+  using memory_space = typename AViewType::memory_space;
+  using Scalar       = typename BViewType::non_const_value_type;
+  using ALayout_t    = typename AViewType::array_layout;
+  using BLayout_t    = typename BViewType::array_layout;
+
+  const int m   = A.extent_int(0);
+  const int n   = A.extent_int(1);
+  const int lda = std::is_same_v<ALayout_t, Kokkos::LayoutRight> ? A.stride(0)
+                                                                 : A.stride(1);
+
+  (void)B;
+
+  const int nrhs = B.extent_int(1);
+  const int ldb  = std::is_same_v<BLayout_t, Kokkos::LayoutRight> ? B.stride(0)
+                                                                 : B.stride(1);
+  int lwork = 0;
+  Kokkos::View<int, memory_space> info("getrf info");
+
+  CudaLapackSingleton& s = CudaLapackSingleton::singleton();
+  KOKKOS_CUSOLVER_SAFE_CALL_IMPL(
+      cusolverDnSetStream(s.handle, space.cuda_stream()));
```

Comment:
@lucbv I just want to make sure I understand. Do we have only one cuSOLVER handle used for all streams or is a new cuSOLVER handle is created every time we call ```CudaLapackSingleton::singleton()```? In the latter case, would it end up having a lot of cuSOLVER handles (until Kokkos is finalized) if the gesv functions are called many times in the application code? Does cuSOLVER have a limit of how many handles should be created? Are there ways to reuse these handles or destroy them when we do not need them anymore?

Reply:
We should have a single handle, which is what we call the singleton and we check if it has already been created.
In that case we just pass it to the user otherwise we create it.
The upside is we only have one actual handle and it get initialized only once.
The downside is that if you use something like std::threads to manage streams, all your std::threads will have the same handle and that is not thread safe...

Reply:
@lucbv Thanks for confirming.  So it is okay to have a single handle for different streams on the same device.  It is not necessary to create a separate handle per stream.